diff --git a/python/mori/shmem/api.py b/python/mori/shmem/api.py
--- a/python/mori/shmem/api.py
+++ b/python/mori/shmem/api.py
@@ -113,7 +113,11 @@ def shmem_torch_process_group_init(group_name: str):
     if rank == 0:
         uid = shmem_get_unique_id()
-        dist.broadcast_object_list([uid], src=0, group=group)
+        dist.broadcast_object_list(
+            [uid], src=dist.get_global_rank(group, 0), group=group
+        )
     else:
         uid_list = [None]
-        dist.broadcast_object_list(uid_list, src=0, group=group)
+        dist.broadcast_object_list(
+            uid_list, src=dist.get_global_rank(group, 0), group=group
+        )
         uid = uid_list[0]
